requirements/progress.py

```diff
@@ -0,0 +1,318 @@
+
+class Threshold(object):
+    def __init__(self, threshold_type, number, criterion):
+        self.type = threshold_type
+        self.cutoff = number
+        self.criterion = criterion
+    def cutoff_for_criterion(self, criterion):
+        if self.criterion == criterion:
+            co = self.cutoff
+        elif self.criterion == CRITERION_SUBJECTS:
+            co = self.cutoff * DEFAULT_UNIT_COUNT
+        else:
+            co = self.cutoff / DEFAULT_UNIT_COUNT
+        return co
+    def get_actual_cutoff(self):
+        if self.type == THRESHOLD_TYPE_GT:
+            return self.cutoff + 1
+        elif self.type == THRESHOLD_TYPE_LT:
+            return self.cutoff - 1
+        return self.cutoff
+    def is_satisfied_by(self, subject_progress, unit_progress):
+        progress = (subject_progress, unit_progress)[self.criterion == CRITERION_UNITS]
+        actualcutoff = self.get_actual_cutoff()
+        if self.type == THRESHOLD_TYPE_LT or self.type == THRESHOLD_TYPE_LTE:
+            return progess <= actualcutoff
+        elif self.type == THRESHOLD_TYPE_GT or self.type == THRESHOLD_TYPE_GTE:
+            return progress >= actualcutoff
+    def __repr__(self):
+        return self.type + " " + self.criterion + " " + self.number
+
+class RequirementsProgress(object):
+    """
+    Stores a user's progress towards a given requirements statement. This object
+    wraps a requirements statement and has a to_json_object() method which
+    returns the statement's own JSON dictionary representation with progress
+    information added.
+
+    Note: This class is maintained separately from the Django model so that
+    persistent information can be stored in a database-friendly format, while
+    information specific to a user's request is transient.
+    """
+    def __init__(self, statement, list_path):
+        """Initializes a progress object with the given requirements statement."""
+        self.statement = statement
+        self.list_path = list_path
+        self.children = []
+        if self.statement.requirements.exists():
+            for index, child in enumerate(self.statement.requirements.all()):
+                self.children.append(RequirementsProgress(child, list_path+"."+str(index)))
+
+    def courses_satisfying_req(self, courses):
+        if self.statement.requirement is not None:
+            return set([c for c in courses if c.satisfies(self.statement.requirement)])
+        return []
+
+
+    def compute(self, courses, progress_overrides):
+        """Computes and stores the status of the requirements statement using the
+        given list of Course objects."""
+        # Compute status of children and then self, adapted from mobile apps'
+        # computeRequirementsStatus method
+        courses = list(set(courses))
+        satisfied_courses = set()
+        current_threshold = Threshold(self.statement.threshold_type, self.statement.threshold_cutoff, self.statement.threshold_criterion)
```

Comment:
Going along with what I said before about the `Threshold` class, it'd be great if you could just use it everywhere in this method instead of looking at `self.statement.threshold_X`.